opt_einsum/backends/shared.py

```diff
@@ -0,0 +1,116 @@
+import contextlib
+import numbers
+from collections import OrderedDict
+
+from ..parser import get_symbol
+from .dispatch import get_func
+
+_SHARING_STACK = []
+_CURRENT_BACKEND = []
+
```

Comment:
Could these just be explicit ``global`` variables? Or does the list functionality enable nested ``sharing`` with different caches or something?